crates/sui-indexer-alt-framework/src/pipeline/concurrent/committer.rs

```diff
@@ -213,3 +213,353 @@ pub(super) fn committer<H: Handler + 'static>(
+mod tests {
+    use std::{
+        collections::HashMap,
+        sync::{Arc, Mutex},
+    };
+
+    use async_trait::async_trait;
+    use sui_types::full_checkpoint_content::CheckpointData;
+    use tokio::sync::mpsc;
+    use tokio_util::sync::CancellationToken;
+
+    use crate::{
+        metrics::IndexerMetrics,
+        pipeline::{
+            concurrent::{BatchedRows, Handler},
+            Processor, WatermarkPart,
+        },
+        store::CommitterWatermark,
+        testing::mock_store::*,
+        FieldCount,
+    };
+
+    use super::*;
+
+    #[derive(Clone, FieldCount)]
+    pub struct StoredData {
+        pub cp_sequence_number: u64,
+        pub tx_sequence_numbers: Vec<u64>,
+        pub commit_failure_remaining: Arc<Mutex<usize>>,
+        pub commit_delay_ms: u64,
+    }
+
+    impl Default for StoredData {
+        fn default() -> Self {
+            Self {
+                cp_sequence_number: 0,
+                tx_sequence_numbers: Vec::new(),
+                commit_failure_remaining: Arc::new(Mutex::new(0)),
+                commit_delay_ms: 0,
+            }
+        }
+    }
+
+    pub struct DataPipeline;
+
+    impl Processor for DataPipeline {
+        const NAME: &'static str = "data";
+
+        type Value = StoredData;
+
+        fn process(&self, _checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
+            Ok(vec![])
+        }
+    }
+
+    #[async_trait]
+    impl Handler for DataPipeline {
+        type Store = MockStore;
+
+        async fn commit<'a>(
+            values: &[StoredData],
+            conn: &mut MockConnection<'a>,
+        ) -> anyhow::Result<usize> {
+            for value in values {
+                // If there's a delay, sleep for that duration
+                if value.commit_delay_ms > 0 {
+                    tokio::time::sleep(Duration::from_millis(value.commit_delay_ms)).await;
+                }
+
+                // If there are remaining failures, fail the commit and decrement the counter
+                {
+                    let mut remaining = value.commit_failure_remaining.lock().unwrap();
+                    if *remaining > 0 {
+                        *remaining -= 1;
+                        return Err(anyhow::anyhow!(
+                            "Commit failed, remaining failures: {}",
+                            *remaining
+                        ));
+                    }
+                }
+
+                // Lock, insert, and immediately drop the lock
+                {
+                    let mut data = conn.0.data.lock().unwrap();
+                    data.insert(value.cp_sequence_number, value.tx_sequence_numbers.clone());
+                }
+            }
+            Ok(values.len())
+        }
+    }
+
+    struct TestSetup {
+        store: MockStore,
+        batch_tx: mpsc::Sender<BatchedRows<DataPipeline>>,
+        watermark_rx: mpsc::Receiver<Vec<WatermarkPart>>,
+        committer_handle: JoinHandle<()>,
+    }
+
```

Comment:
```suggestion

/// Creates and spawns a committer task with the provided mock store, along with
/// all necessary channels and configuration. The committer runs in the background
/// and can be interacted with through the returned channels.
```

Quick comment so we can build context on future revisits

Reply:
Sounds good. Done.